src/player.cpp

```diff
@@ -11414,7 +11417,12 @@ bool player::can_sleep()
         sleepy += int((fatigue - 192) / 16);
     }
     sleepy += rng(-8, 8);
-    sleepy -= 2 * stim;
+    if( !has_trait("INSOMNIA") ) {
+        sleepy -= 2 * stim;
+    } else {
+        // Make it harder for insomniac to get around the trait
+        sleepy -= stim;
+    }
```

Comment:
Doesn't that also incidentally make positive stim less of an issue when trying to sleep as insomniac?
Perhaps
if( !has_trait("INSOMNIA") || ( stim > 0 ) ) {

would work better?